# Cyclistic Ride-Share Data Exploration & Validation

## Introduction

Exploration and validation of 2024 Cyclistic ride-share data to assess data quality, validate schema consistency, and explore rider behavior patterns throughout the development of the Power BI dashboard.

### Environment Setup

Import required libraries for data exploration and validation.

In [82]:
import numpy as np
import pandas as pd

The required libraries are available for data exploration and validation tasks.

### Data Loading

Load the combined Cyclistic trip dataset for exploratory analysis and validation.

In [83]:
dataframe=pd.read_csv(r"C:\Users\saswa\Desktop\DA projects\Python Projects\DA case study cyclistic project\All the data\Mysql upload datasets\combined_tripdata(query--export results).csv")

The dataset has been loaded successfully and is ready for initial inspection.

### Initial Data Inspection

Preview the dataset structure and verify that the data has been loaded correctly.

In [84]:
dataframe.head()

,ride_id,rideable_type,started_at,ended_at,start_station_name,start_station_id,end_station_name,end_station_id,start_lat,start_lng,end_lat,end_lng,member_casual,ride_length,day_of_week,ride_length_minutes
0,000019B6351D28DF,electric_bike,2024-01-12 01:14:51,2024-01-12 01:19:56.000000,MAY ST & TAYLOR ST,13160,RACINE AVE & 15TH ST,13304,41.869421,-87.655420,41.861267,-87.656625,member,00:05:05,6,5.08
1,0000F0DC283B1EF1,classic_bike,2024-01-26 11:51:01,2024-01-26 12:33:37.000000,CLARENDON AVE & JUNIOR TER,13389,STATE ST & VAN BUREN ST,TA1305000035,41.961004,-87.649603,41.877181,-87.627844,member,00:42:36,6,42.60
2,00013D27FFAD1F38,classic_bike,2024-01-25 09:42:40,2024-01-25 09:57:19.000000,CLARK ST & COLUMBIA AVE,RP-008,WESTERN AVE & GRANVILLE AVE,KA1504000129,42.004451,-87.672402,41.994161,-87.689438,casual,00:14:39,5,14.65
3,0001941A2BC4AFB7,classic_bike,2024-01-25 17:29:06,2024-01-25 18:05:03.000000,DUSABLE LAKE SHORE DR & MONROE ST,13300,LAKE PARK AVE & 47TH ST,TA1308000035,41.880958,-87.616743,41.809443,-87.591875,member,00:35:57,5,35.95
4,0002D4A03DCA48D9,electric_bike,2024-01-27 17:43:50,2024-01-27 17:47:23.000000,CLARK ST & ELM ST,TA1307000039,WELLS ST & HURON ST,TA1306000012,41.902764,-87.631655,41.894722,-87.634362,member,00:03:33,7,3.55


The first few records provide an initial view of the dataset and confirm that the data has been loaded successfully.

### Dataset Information

Inspect the dataset structure, column data types, and non-null value counts.

In [85]:
dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ride_id              1000 non-null   object 
 1   rideable_type        1000 non-null   object 
 2   started_at           1000 non-null   object 
 3   ended_at             1000 non-null   object 
 4   start_station_name   1000 non-null   object 
 5   start_station_id     1000 non-null   object 
 6   end_station_name     1000 non-null   object 
 7   end_station_id       1000 non-null   object 
 8   start_lat            1000 non-null   float64
 9   start_lng            1000 non-null   float64
 10  end_lat              1000 non-null   float64
 11  end_lng              1000 non-null   float64
 12  member_casual        1000 non-null   object 
 13  ride_length          1000 non-null   object 
 14  day_of_week          1000 non-null   int64  
 15  ride_length_minutes  1000 non-null   fl

The dataset schema confirms the presence of 16 columns with complete non-null counts across the sampled records. The next step is to convert date and time fields into appropriate datetime formats for further exploration.

### Datetime Conversion

Convert date and time columns to datetime format for data validation and exploration.

In [86]:
dataframe['started_at'] = pd.to_datetime(dataframe['started_at'])

The datetime conversion has been applied successfully. The next step is to validate temporal coverage and confirm that records are present across the expected time period.

### Temporal Coverage Validation

Validate the temporal coverage of the sampled dataset and confirm the month represented in the extracted records.

In [87]:
dataframe.groupby(dataframe['started_at'].dt.month).size()

started_at
1    1000
dtype: int64

The sample dataset review confirms that the extracted records belong to January 2024. The next step is to load all monthly Cyclistic trip datasets and validate consistency across the complete 2024 analysis period.

### Monthly Dataset Loading

Load all monthly Cyclistic trip datasets for 2024 to validate data consistency across the complete analysis period.

In [88]:
df_202401 = pd.read_csv(r"C:\Users\saswa\Desktop\DA projects\Python Projects\DA case study cyclistic project\Power BI Dashboard\Data\202401-divvy-tripdata.csv")

df_202402 = pd.read_csv(r"C:\Users\saswa\Desktop\DA projects\Python Projects\DA case study cyclistic project\Power BI Dashboard\Data\202402-divvy-tripdata.csv")

df_202403 = pd.read_csv(r"C:\Users\saswa\Desktop\DA projects\Python Projects\DA case study cyclistic project\Power BI Dashboard\Data\202403-divvy-tripdata.csv")

df_202404 = pd.read_csv(r"C:\Users\saswa\Desktop\DA projects\Python Projects\DA case study cyclistic project\Power BI Dashboard\Data\202404-divvy-tripdata.csv")

df_202405 = pd.read_csv(r"C:\Users\saswa\Desktop\DA projects\Python Projects\DA case study cyclistic project\Power BI Dashboard\Data\202405-divvy-tripdata.csv")

df_202406 = pd.read_csv(r"C:\Users\saswa\Desktop\DA projects\Python Projects\DA case study cyclistic project\Power BI Dashboard\Data\202406-divvy-tripdata.csv")

df_202407 = pd.read_csv(r"C:\Users\saswa\Desktop\DA projects\Python Projects\DA case study cyclistic project\Power BI Dashboard\Data\202407-divvy-tripdata.csv")

df_202408 = pd.read_csv(r"C:\Users\saswa\Desktop\DA projects\Python Projects\DA case study cyclistic project\Power BI Dashboard\Data\202408-divvy-tripdata.csv")

df_202409 = pd.read_csv(r"C:\Users\saswa\Desktop\DA projects\Python Projects\DA case study cyclistic project\Power BI Dashboard\Data\202409-divvy-tripdata.csv")

df_202410 = pd.read_csv(r"C:\Users\saswa\Desktop\DA projects\Python Projects\DA case study cyclistic project\Power BI Dashboard\Data\202410-divvy-tripdata.csv")

df_202411 = pd.read_csv(r"C:\Users\saswa\Desktop\DA projects\Python Projects\DA case study cyclistic project\Power BI Dashboard\Data\202411-divvy-tripdata.csv")

df_202412 = pd.read_csv(r"C:\Users\saswa\Desktop\DA projects\Python Projects\DA case study cyclistic project\Power BI Dashboard\Data\202412-divvy-tripdata.csv")

The monthly datasets have been loaded successfully. The next step is to verify schema consistency across all files before combining them for further exploration.

### Schema Consistency Validation

Compare column structures across all monthly datasets to verify schema consistency prior to data consolidation.

In [89]:
datasets = {
    "202401": df_202401,
    "202402": df_202402,
    "202403": df_202403,
    "202404": df_202404,
    "202405": df_202405,
    "202406": df_202406,
    "202407": df_202407,
    "202408": df_202408,
    "202409": df_202409,
    "202410": df_202410,
    "202411": df_202411,
    "202412": df_202412
}

for name, df in datasets.items():
    print(f"\n{name}")
    print(df.columns.tolist())


202401
['ride_id', 'rideable_type', 'started_at', 'ended_at', 'start_station_name', 'start_station_id', 'end_station_name', 'end_station_id', 'start_lat', 'start_lng', 'end_lat', 'end_lng', 'member_casual']

202402
['ride_id', 'rideable_type', 'started_at', 'ended_at', 'start_station_name', 'start_station_id', 'end_station_name', 'end_station_id', 'start_lat', 'start_lng', 'end_lat', 'end_lng', 'member_casual']

202403
['ride_id', 'rideable_type', 'started_at', 'ended_at', 'start_station_name', 'start_station_id', 'end_station_name', 'end_station_id', 'start_lat', 'start_lng', 'end_lat', 'end_lng', 'member_casual']

202404
['ride_id', 'rideable_type', 'started_at', 'ended_at', 'start_station_name', 'start_station_id', 'end_station_name', 'end_station_id', 'start_lat', 'start_lng', 'end_lat', 'end_lng', 'member_casual']

202405
['ride_id', 'rideable_type', 'started_at', 'ended_at', 'start_station_name', 'start_station_id', 'end_station_name', 'end_station_id', 'start_lat', 'start_lng',

The column structures are consistent across all monthly datasets, confirming schema compatibility for consolidation. 
The next step is to consolidate the monthly files into a single dataset for unified validation and exploration.

### Dataset Consolidation

Combine all monthly Cyclistic trip datasets into a single dataframe for comprehensive validation and exploratory analysis.

In [90]:
combined_tripdata = pd.concat([
    df_202401,
    df_202402,
    df_202403,
    df_202404,
    df_202405,
    df_202406,
    df_202407,
    df_202408,
    df_202409,
    df_202410,
    df_202411,
    df_202412
], ignore_index=True)

The monthly datasets have been successfully consolidated into a single dataframe. The next step is to inspect the structure, data types, and completeness of the combined dataset.

### Combined Dataset Structure Validation

Inspect the consolidated dataset structure, column data types, and memory usage after merging all monthly files.

In [91]:
combined_tripdata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5860568 entries, 0 to 5860567
Data columns (total 13 columns):
 #   Column              Dtype  
---  ------              -----  
 0   ride_id             object 
 1   rideable_type       object 
 2   started_at          object 
 3   ended_at            object 
 4   start_station_name  object 
 5   start_station_id    object 
 6   end_station_name    object 
 7   end_station_id      object 
 8   start_lat           float64
 9   start_lng           float64
 10  end_lat             float64
 11  end_lng             float64
 12  member_casual       object 
dtypes: float64(4), object(9)
memory usage: 581.3+ MB


The consolidated dataset contains over 5.8 million records and 13 columns with a consistent schema across all monthly files.
The next step is to assess missing values in key station-related fields and evaluate the completeness of location information required for ride location analysis.

### Missing Value Assessment

Evaluate missing values in start and end station names to validate data completeness and identify potential gaps in station-level records.

In [92]:
print("\nNull values:\n")
print(combined_tripdata[['start_station_name', 'end_station_name']].isnull().sum())


Null values:

start_station_name    1073951
end_station_name      1104653
dtype: int64


The station name fields contain a notable number of missing values. The next step is to verify whether additional records contain blank strings that may not be captured by the null value assessment.

### Blank String Validation

Check for blank string entries in station-related fields to identify missing information that is not represented as null values.

In [93]:
print("\nBlank string counts:\n")

start_blank = (
    combined_tripdata['start_station_name']
    .astype(str)
    .str.strip()
    .eq('')
    .sum()
)

end_blank = (
    combined_tripdata['end_station_name']
    .astype(str)
    .str.strip()
    .eq('')
    .sum()
)

print("Blank start_station_name:", start_blank)
print("Blank end_station_name:", end_blank)


Blank string counts:

Blank start_station_name: 0
Blank end_station_name: 0


The blank string assessment indicates that missing station information is represented as null values rather than empty text entries. The next step is to quantify the total number of problematic station records by combining both null and blank-value checks.

### Problematic Station Record Quantification

Quantify records with missing or blank station names to assess the extent of station-related data quality issues.

In [94]:
problematic_start = (
    combined_tripdata['start_station_name'].isnull()
    |
    combined_tripdata['start_station_name']
    .astype(str)
    .str.strip()
    .eq('')
)

problematic_end = (
    combined_tripdata['end_station_name'].isnull()
    |
    combined_tripdata['end_station_name']
    .astype(str)
    .str.strip()
    .eq('')
)

print("\nTotal problematic start stations:", problematic_start.sum())
print("Total problematic end stations:", problematic_end.sum())


Total problematic start stations: 1073951
Total problematic end stations: 1104653


The overall count of problematic station records has been quantified. The next step is to examine specific missing-value patterns to determine whether station information is missing independently or across both fields simultaneously.

### Missing Pattern Analysis

Assess whether missing station information occurs independently in start or end station fields, or simultaneously in both fields.

In [95]:
# Only start station missing
only_start_missing = (
    combined_tripdata['start_station_name'].isnull()
    &
    combined_tripdata['end_station_name'].notnull()
).sum()

In [96]:
# Only end station missing
only_end_missing = (
    combined_tripdata['start_station_name'].notnull()
    &
    combined_tripdata['end_station_name'].isnull()
).sum()

In [97]:
both_missing = (
    combined_tripdata['start_station_name'].isnull()
    &
    combined_tripdata['end_station_name'].isnull()
).sum()

In [98]:
print("Only start missing:", only_start_missing)
print("Only end missing:", only_end_missing)
print("Both missing:", both_missing)

Only start missing: 547606
Only end missing: 578308
Both missing: 526345


The missing pattern analysis shows that station information can be absent independently in either field or simultaneously in both fields. The next step is to further validate the nature of these missing values by distinguishing between null values, empty strings, and whitespace-only entries.

### Missing Value Type Validation

Validate whether missing station information is represented as null values, empty strings, or whitespace-only entries.

In [99]:
# Null values
null_count = combined_tripdata['start_station_name'].isnull().sum()

# Empty strings only
empty_count = (
    combined_tripdata['start_station_name']
    .astype(str)
    .eq('')
).sum()

# Whitespace-only strings
whitespace_count = (
    combined_tripdata['start_station_name']
    .astype(str)
    .str.match(r'^\s+$', na=False)
).sum()

print("Null values:", null_count)
print("Empty strings:", empty_count)
print("Whitespace-only strings:", whitespace_count)

Null values: 1073951
Empty strings: 0
Whitespace-only strings: 0


The validation confirms that missing station information is represented exclusively as null values, with no empty or whitespace-only entries detected. The next step is to examine ride frequency across stations to identify low-activity locations and validate station-level data distribution.

### Station Usage Distribution Analysis

Analyze ride counts by start station to identify low-frequency stations and assess the distribution of station-level activity.

In [100]:
# Remove null start stations first
clean_stations = combined_tripdata.dropna(subset=['start_station_name'])

# Count rides per start station
station_counts = (
    clean_stations['start_station_name']
    .value_counts()
    .sort_values(ascending=True)
)

# Show actual bottom 10
print(station_counts.head(10))

start_station_name
Public Rack - Oglesby Ave & 105th St          1
Public Rack - Torrence Ave & 103rd St         1
Public Rack - Ave L & 105th St                1
Public Rack - 53rd St & Indiana Ave           1
Public Rack - Lafayette Ave & 87th St SW      1
Public Rack - Kedzie Ave & 83rd St            1
Public Rack - Homan Ave & 111th St            1
Public Rack - Vincennes Ave & 99th St         1
Public Rack - California Ave & Pershing Rd    1
Public Rack - Roscoe St & Osceola Ave         1
Name: count, dtype: int64


The station usage distribution highlights several low-frequency stations within the dataset. The next step is to determine the total number of unique start stations represented in the consolidated dataset.

### Unique Station Count Assessment

Determine the total number of unique start stations represented in the consolidated dataset.

In [101]:
# combined_tripdata['start_station_name'].value_counts()
combined_tripdata['start_station_name'].nunique()

1808

The dataset contains 1,808 unique start stations, indicating broad geographic coverage across the bike-share network. The next step is to examine ride activity at the upper end of the distribution by identifying the most frequently used stations.

## Conclusion

The exploratory analysis and validation process confirmed schema consistency across all monthly datasets and identified the primary data quality considerations within the 2024 Cyclistic ride-share data.

Key validation findings included:

- Consistent column structures across all monthly files
- Missing station information represented as null values rather than blank or whitespace entries
- Broad station coverage across the network with over 1,800 unique start stations
- Successful consolidation of all monthly datasets into a unified analysis dataset

These validation and exploration steps provided the foundation for the Power BI dashboard and subsequent rider behavior analysis.